csharp/src/ProtocolBuffers/DescriptorProtos/PartialClasses.cs

```diff
@@ -30,36 +30,18 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endregion
 
-// This file just contains partial classes for each of the
-// autogenerated classes, so that they implement
-// IDescriptorProto
+// This file just contains partial classes for any autogenerated classes that need additional support.
 namespace Google.Protobuf.DescriptorProtos
 {
-    public partial class DescriptorProto : IDescriptorProto<MessageOptions>
-    {
-    }
-
-    public partial class EnumDescriptorProto : IDescriptorProto<EnumOptions>
-    {
-    }
-
-    public partial class EnumValueDescriptorProto : IDescriptorProto<EnumValueOptions>
-    {
-    }
-
-    public partial class FieldDescriptorProto : IDescriptorProto<FieldOptions>
-    {
-    }
-
-    public partial class FileDescriptorProto : IDescriptorProto<FileOptions>
-    {
-    }
-
-    public partial class MethodDescriptorProto : IDescriptorProto<MethodOptions>
-    {
-    }
-
-    public partial class ServiceDescriptorProto : IDescriptorProto<ServiceOptions>
+    internal partial class FieldDescriptorProto
     {
+        // We can't tell the difference between "explicitly set to 0" and "not set"
+        // in proto3, but we need to tell the difference for OneofIndex. descriptor.proto
+        // is really a proto2 file, but the runtime doesn't know about proto2 semantics...
+        // We fake it by defaulting to -1.
+        partial void OnConstruction()
+        {
+            OneofIndex = -1;

```

Comment:
Is this the only reason why we need to generate OnConstruction everywhere? If so, that's a bit unfortunate. Perhaps we could at least add a TODO to try to get rid of this later?